### PR TITLE
Removed OrderedDict from io.ascii

### LINKED ISSUE #10888: Remove OrderedDict for astropy 4.2
### Description

Based on https://github.com/astropy/astropy-APEs/pull/63, it looks like astropy 4.2 will support only Python 3.7+, so we should be able to count on the built-in `dict` being ordered. 

So can we get rid of all the `OrderedDict` now?

### Additional context

https://mail.python.org/pipermail/python-dev/2017-December/151283.html

### PR DESCRIPTION
**Pull request by alexandrecesarba**
### Description

This pull request removes OrderedDict from files in io.ascii.

The files are: core.py , daophot.py , ecsv.py , fastbasic.py , ipac.py and test_read.py. The OrderedDicts are normal dictionaries now.

Fixes part of #10888

### CODE REVIEW

**Comment by dhomeier** [comment]
How was this not caught by codestyle checks?

**Comment by pllim** [comment]
I think this might be a tricky one. It causes real failures in cosmology tests. I am turning this into draft for now.

**Comment by alexandrecesarba** [comment]
pre-commit.ci autofix

**Comment by dhomeier** [comment]
Pinging @nstarman – for some reason this is causing the `meta` of cosmologies to be sorted alphabetically on reading back in (or already on writing), though for roundtripping ECSV directly everything passes. Something happening in `cosmology._io.mapping`?

**Comment by dhomeier** [comment]
So yaml is reordering a `dict`, but not `OrderedDict`?

And I have been wondering as well if e.g. https://github.com/astropy/astropy/issues/10888#issuecomment-721852976 might be relevant – sounds like something like this could be happening within `mapping`.

**Comment by mhvk** [comment]
Yes, `yaml` dict does not preserve order. And, indeed, #10888 is the issue I was thinking of, with the PRs that link to it (including those closed without merging).

**Comment by dhomeier** [comment]
But `yaml.dump` has a `sort_keys` option. So presumably setting `sort_keys=False` in
https://github.com/astropy/astropy/blob/bbdf2ce195ddd6ed88179c21b0af93130c8ffe2f/astropy/table/meta.py#L387-L389
would fix this problem (unless there are any header structures where sorting is explicitly desired and expected, but I doubt that). At least testing with plain `AstropyDumper` indicates so (`yaml.load` seems to preserve order already by default).

However that would link this PR to a _Remove OrderedDict from `table`_ task along with several dozen other instances in that module.

**Comment by taldcroft** [comment]
@dhomeier @alexandrecesarba - if we make the suggested change in `meta.py`, do tests pass?

**Comment by alexandrecesarba** [comment]
pre-commit.ci autofix

**Comment by dhomeier** [comment]
OK, looks like
https://github.com/astropy/astropy/blob/bbdf2ce195ddd6ed88179c21b0af93130c8ffe2f/astropy/cosmology/_io/yaml.py#L203
is directly calling `io.misc.yaml`'s own version of the dumper. Would need @nstarman's input whether it makes sense to set `sort_keys=False` there as well.
We could also add it more widely as
`kwargs.setdefault("sort_keys", False)`
to
https://github.com/astropy/astropy/blob/bbdf2ce195ddd6ed88179c21b0af93130c8ffe2f/astropy/io/misc/yaml.py#L379-L381

Could you try the latter change to see if that lets the tests pass?

File: astropy/io/misc/yaml.py [hunk 1]
```diff
@@ -377,5 +377,5 @@ def dump(data, stream=None, **kwargs):
 
     """
     kwargs["Dumper"] = AstropyDumper
-    kwargs.setdefault("default_flow_style", None)
+    kwargs.setdefault("sort_keys", False)
```
Comment (dhomeier) [inline astropy/io/misc/yaml.py:380]: Why did you remove the `default_flow_style` default? That's an unrelated setting, although it's unclear to me if this has now made the tests still fail...

Comment (alexandrecesarba) [reply astropy/io/misc/yaml.py:380]: Sorry, that wasn't supposed to happen. Thanks for noticing, I'll make the correct changes now as you suggested.

File: docs/changes/io.ascii/15496.api.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Removed usage of ``OrderedDict`` from several files within the ``io.ascii`` subpackage. Affected files include: ``core.py``, ``daophot.py``, ``ecsv.py``, ``fastbasic.py``, ``ipac.py``, and ``test_read.py``. With the guaranteed insertion order in dictionaries from Python 3.7 onwards, the shift to standard dictionaries simplifies the code and reduces reliance on specialized data structures. This refactoring aligns with issue #10888.
```
Comment (dhomeier) [inline docs/changes/io.ascii/15496.api.rst:1]:
```suggestion
Replaced ``OrderedDict`` instances within ``io.ascii`` with standard dictionaries, which preserve insertion order as of Python 3.7. Changed calls to ``yaml.dump`` to default to ``sort_keys=False`` to ensure this order is preserved on serialization as well.
```
No need to list all modified files here, as they will be easily identifiable by `git log --name-status`. Internal uses of dict/OrderedDict may not require a changelog at all, but if this finally succeeds with the modified dump, it should probably have one, as it will output different yaml files for standard dictionaries.

@pllim : the other PRs for `io` were merged w/o changelog, but #15491 has one for API change; not quite sure if this really constitutes a change in API (see above).

Comment (alexandrecesarba) [reply docs/changes/io.ascii/15496.api.rst]: Thanks for your advice and patience!

Comment (dhomeier) [reply docs/changes/io.ascii/15496.api.rst]: Thank you for your persistence on what turns out to be a far from straightforward refactor!
I just committed the change directly since the tests had been stuck on the previous commit somehow; unfortunately we are still not quite there.

**Comment by dhomeier** [comment]
There are some additional failures in the doctests now, but those actually show that the `yaml.dump` change actually works as expected – the docs just still expect an output that is reordered alphabetically.

However the real issue in `cosmology.io` seems unrelated to this, as that test is in fact going through the ECSV writer. Just realising though that it only fails for the `_subclass_partial_info` tests, and those are actually reading back a manipulated file that is opened and written back as a `QTable`. So I think we are back to having this linked to `OrderedDict` still being used in `table` (and I suspect it is not just `QTable`; notably `TableColumns` still is an `OrderedDict` subclass).

So I think this will still have to be synchronised with removing `OrderedDict` from `table` as well.

**Comment by nstarman** [comment]
LMK what I can do to help on the cosmology side. Though if this is a Table issue, then perhaps the cosmology tests will "just work" when Table is addressed?

**Comment by nstarman** [comment]
> Would need @nstarman's input whether it makes sense to set sort_keys=False there as well.

Key order is important in cosmology ONLY for doctests. 
I think having deterministic tests is somewhat important, so it would be good if the YAML dump preserved ordering of passed an OrderedDict. 

Some logic like
```
if isinstance(sort_keys, bool):
    pass
elif sort_keys is None:
    sort_keys = True if isinstance(obj, OrderedDict) else False
else:
    raise TypeError("what are you doing?")
```

**Comment by dhomeier** [comment]
> Key order is important in cosmology ONLY for doctests.

In general this only seems to be an issue with some doctests. So if we preserve order in the YAML output of `dict` in general, as I think we should (see below), just need to update the expected output in those docs.

Cosmology tests like `*_readwrite_ecsv_*` however do expect key order to be preserved as well, as checks like
`assert got.meta == cosmo.meta` will fail on changed key orders _as long as the `meta` are `OrderedDict`s_.
If both are standard `dict`, they will still qualify as equal, even though any given `dict` is preserving its key order.
So the issue may indeed resolve itself once everything is changed to `dict` (which supposedly means once `Table.meta` is becoming a regular `dict`, since `cosmology` is not explicitly calling `OrderedDict` anywhere – in particular as the one failing test group `test_readwrite_ecsv_subclass_partial_info` appears to be failing because of the intermediate `QTable` read/write operation).

> I think having deterministic tests is somewhat important, so it would be good if the YAML dump preserved ordering of passed an OrderedDict, which implies the person cares about ordering.

Thought for a moment I misunderstood https://github.com/astropy/astropy/pull/15496#issuecomment-1767312808,  but that is already the default for `OrderedDict`, if you e.g. directly compare the output of `yaml.dump(Cosmo.meta)` and `yaml.dump(dict(Cosmo.meta))`.
But the ultimate point of #10888 is to replace `OrderedDict` everywhere, which implies that we should always care about ordering in `dict` as well. Thus my suggestion to have `io.misc.yaml.dump` preserve order  by default.
Although we could still make this input-dependent along the lines of your suggestion as
`sort_keys = False if isinstance(obj, (dict, OrderedDict)) else True`
if there are really other objects where we want to keep re-sorting enabled.

But it looks like all of this really depends on getting `table` refactored as well.

**Comment by MridulS** [comment]
@alexandrecesarba I started looking into this while going through https://github.com/astropy/astropy/issues/10888, do you still have the bandwidth to finish this up? I can cherry pick the commits and start a new PR if you aren't available right now :)

**Comment by MridulS** [comment]
I started looking a bit more into this and we just need to do change the default factory in `MetaData` to `dict` to make cosmology tests happy. I will create a new PR which builds on top of this :)

``` diff
--- a/astropy/utils/metadata/core.py
+++ b/astropy/utils/metadata/core.py
@@ -79,7 +79,7 @@ class MetaData:
         None
     """
 
-    def __init__(self, doc="", copy=True, *, default_factory=OrderedDict):
+    def __init__(self, doc="", copy=True, *, default_factory=dict):
```

**Comment by dhomeier** [comment]
@MridulS, @taldcroft, is #16250 superseding this PR as well?
